cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ProgramNotificationSubscriberService.java

```diff
@@ -361,7 +395,34 @@ private void handleProgramEvent(ProgramRunId programRunId, ProgramRunStatus prog
         appMetadataStore.recordProgramDeprovisioning(programRunId, messageIdBytes);
         return Optional.of(provisioningService.deprovision(programRunId, datasetContext));
       case DEPROVISIONED:
-        appMetadataStore.recordProgramDeprovisioned(programRunId, endTs, messageIdBytes);
+        RunRecordMeta meta = appMetadataStore.recordProgramDeprovisioned(programRunId, endTs, messageIdBytes);
+        if (meta == null || meta.getCluster() == null || meta.getCluster().getNumNodes() == null) {
+          // this should not happen since when the cluster info should be there after deprovision
+          return Optional.empty();
+        }
+        Optional<ProfileId> profile = SystemArguments.getProfileIdFromArgs(programRunId.getNamespaceId(),
+                                                                           meta.getSystemArgs());
+
+        if (!profile.isPresent()) {
+          // Each program run should be associated with a profile, so this should never happen
+          LOG.warn("Will not emit metrics for program {} without profile assigned to it, {}",
+                   programRunId, notification);
+          return Optional.empty();
+        }
+
+        ProfileId profileId = profile.get();
+        Integer node = meta.getCluster().getNumNodes();
+        // the node number will be null or 0 in local mode, set it to 1 so that we will just emit the difference
+        // between endts and starts
+        int nodeNum = node == null || node == 0 ? 1 : node;
+
+        // emit the metrics information, increment the count for end state program runs
+        MetricsContext metricsContext = getMetricsContextFroProfile(programRunId, profileId);
+        metricsContext.increment(getMetricName(meta.getStatus()), 1L);
+
+        // node minutes = (end ts of the cluster - start ts) * nodeNum / 60
+        metricsContext.gauge(Constants.Metrics.Program.PROGRAM_NODE_MINUTES,
```

Comment:
I put it here since this class is used for the program and cluster status update, and here we can get the most accurate metrics for the node hours, and also we can put all the metrics update together instead of having some of them elsewhere. I think we can periodically increment this metric in RuntimeMonitor, and if the run finishes, we can gauge here to get a final value.
And from my understanding of metrics, since the metrics context we get uses program run id and profile id, so each program run will has its metric value gauged, so query on a program will return the sum of this metrics. 
But the way we generate the metrics context still bothers me. For now, the namespace tag we uses is the namespace id of the program id, the profile tag is the profileId.toString(). The correct way to me is to have ns tag be the namespace of the profile, and profile tag should be just the profile name. But doing this we are not able to specify where the program is since program ns and profile ns can be different if profile is in system namespace. I think we can discuss this on Monday on exactly how we want to aggregate on these metrics on Monday since that is easy code change but very hard to understand.

Reply:
The problem is that scoped name is not unique to identify the profile. Profiles can have same name but belong to different namespace, they will have same scoped name. If someone use the instance level metadata or metrics api for it, the result will contain entities from these two namespaces, which is very confusing.

Reply:
It's fine as long as the namespace is also part of the context. It's the same thing as app name not being globally unique, but we always use both namespace and app in the context